### PR TITLE
config: Adding config-validation interface, extension and CDS validator

### PR DESCRIPTION
**Pull request by adisuissa**
Commit Message: Adding custom config-validation interface, extension, and a CDS validator.
Additional Description:
Adding a config-validation interface, an extension, that runs per-xDS validations which verify that the config is valid from a business-logic perspective.
This PR also includes a CDS config validator which ensures that the number of clusters doesn't decrease below a threshold.

Risk Level: Med - config plane changes that should not disrupt current config processing
Testing: Added extension unit tests and integration tests.
Docs Changes: Added.
Release Notes: Added.
Platform Specific Features: N/A.
Part of the work on #19855

Signed-off-by: Adi Suissa-Peleg <adip@google.com>

### CODE REVIEW

File: api/envoy/config/core/v3/config_source.proto [hunk 1]
```diff
@@ -149,10 +150,28 @@ message RateLimitSettings {
...
 // filesystem or from an xDS API source. Filesystem configs are watched with
 // inotify for updates.
-// [#next-free-field: 8]
+// [#next-free-field: 9]
 message ConfigSource {
   option (udpa.annotations.versioning).previous_message_type = "envoy.api.v2.core.ConfigSource";
 
+  // This is the same as TypedExtensionConfig. However, extensions.proto cannot be
+  // imported as it defines a message (ExtensionConfigSource) that depends on ConfigSource.
+  // [#next-major-version: refactor the file envoy/config/core/v3/extension.proto
```
Comment (htuch) [inline api/envoy/config/core/v3/config_source.proto:159]: @adisuissa you are free to move protos between files in the same directory/package. This does not change any generated code or wire level artifacts. It might break folks code-wise during import, but not beyond include paths, and we don't have a strong version policy on that. I think it would be the lesser evil in this case.

File: envoy/config/config_validator.h [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+  /**
+   * Validates a given set of resources matching an Incremental update.
+   * @param server A server instance to fetch the state before applying the config.
+   * @param added_resources A list of decoded resources to add to the current state.
+   * @param removed_resources A list of resources to remove from the current state.
+   * @throw optional EnvoyException if the config should be rejected.
+   * @return whether the config is valid or not.
+   */
+  virtual bool validate(Server::Instance& server,
```
Comment (htuch) [inline envoy/config/config_validator.h:47]: The server instance is pretty wide, is there a more useful abstraction to be able to understand a particular xDS config that we can offer? The answer might be "no" and maybe we should offer some libraries to help folks take a `Server` and get to this state. The anti-pattern I worry about is that validators might get quite complicated just in the part that needs to understand current state.

File: source/common/upstream/cds_api_impl.cc [hunk 3]
```diff
@@ -59,8 +63,19 @@ void CdsApiImpl::onConfigUpdate(const std::vector<Config::DecodedResourceRef>& r
 void CdsApiImpl::onConfigUpdate(const std::vector<Config::DecodedResourceRef>& added_resources,
                                 const Protobuf::RepeatedPtrField<std::string>& removed_resources,
                                 const std::string& system_version_info) {
-  auto exception_msgs =
-      helper_.onConfigUpdate(added_resources, removed_resources, system_version_info);
+  std::vector<std::string> exception_msgs;
+  // Run the external config validators (if any).
+  TRY_ASSERT_MAIN_THREAD { executeValidators(added_resources, removed_resources); }
```
Comment (htuch) [inline source/common/upstream/cds_api_impl.cc:68]: I can't remember, but did we discuss doing this in `GrpcSubscriptionImpl::onConfigUpdate`?

File: api/envoy/config/core/v3/config_source.proto (hunk 1 above)
Comment (adisuissa) [reply api/envoy/config/core/v3/config_source.proto:159]: I was under the impression that it changes the Jave file name (`option java_outer_classname = "ConfigSourceProto";` will not hold now), but if that's not important, I agree with you that splitting is the right thing to do.

File: source/common/upstream/cds_api_impl.cc (hunk 3 above)
Comment (adisuissa) [reply source/common/upstream/cds_api_impl.cc:68]: Yes, as we discussed offline, this should be moved to the `onConfigUpdate` that is being called on the singe update message containing all the resources of the same type.

File: envoy/config/config_validator.h (hunk 2 above)
Comment (adisuissa) [reply envoy/config/config_validator.h:47]: The first version, as was specified in the design doc, defined a new struct that included the ClusterManager and ListenerManager. Passing `Server::Instance` is the most generic thing to pass, and I agree that it makes "everything" accessible, which we might want to avoid.
I guess we can start with the struct that has only the cluster and listener managers, and add other things if needed in the future (e.g., for validation of SDS, or RTDS).

File: api/envoy/config/core/v3/config_source.proto (hunk 1 above)
Comment (htuch) [reply api/envoy/config/core/v3/config_source.proto:159]: You're probably right, nonetheless we don't explicitly rule this out in https://github.com/envoyproxy/envoy/blob/main/api/API_VERSIONING.md and have historically done this without any major complaints. If you want to lock this down, as in principle it is inconsistent with the "don't break consuming code" notion, that's fine, but then update style guide. I'm very much on the fence tilting towards not doing that, since there is no problem flagged to fix.

**Comment by adisuissa** [comment]
@htuch this is the latest version, where the main change is moving the external config validator to the GrpcMux/WatchMap level.

This PR is pretty large at the moment (more than half the code is tests, and plumbing), but if you think the design and core is ok, then I can start chopping it to smaller PRs:
1. Refactor config_source.proto and move `ExtensionConfigSource` to a separate file.
2. Introduce the extension interface (no plumbing).
3. Add plumbing.
4. Add the CDS extension.

Let me know what you think.

**Comment by htuch** [comment]
@adisuissa are some pushes missing? You mentoin that you have moved `ExtensionConfigSource` but this seems the same as before..

**Comment by adisuissa** [comment]
> @adisuissa are some pushes missing? You mentoin that you have moved `ExtensionConfigSource` but this seems the same as before..

@htuch Sorry, you are correct. Added the relevant commits, and should be good to go, thanks!

File: envoy/upstream/cluster_manager.h [hunk 4]
```diff
@@ -248,28 +254,32 @@ class ClusterManager {
...
         return warming_cluster->second;
       }
       return absl::nullopt;
     }
 
     ClusterInfoMap active_clusters_;
     ClusterInfoMap warming_clusters_;
+
+    // Number of clusters that were dynamically added via API (xDS). This will be
+    // less than or equal to the number of `active_clusters_` and `warming_clusters_`.
```
Comment (htuch) [inline envoy/upstream/cluster_manager.h:273]: We could add some `ASSERT`s here as well.

File: source/common/config/external_config_validators.h [hunk 5]
```diff
@@ -0,0 +1,44 @@
+#pragma once
+
+#include "envoy/config/config_validator.h"
+
+namespace Envoy {
+namespace Config {
+
+// Represents a collection of external config validators for all xDS
```
Comment (htuch) [inline source/common/config/external_config_validators.h:8]: What is "external" here?

File: source/common/config/external_config_validators_impl.cc [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+void ExternalConfigValidatorsImpl::executeValidators(
+    absl::string_view type_url, const std::vector<DecodedResourcePtr>& resources) {
+  auto validators_it = validators_map_.find(type_url);
+  if (validators_it != validators_map_.end()) {
+    auto& validators = validators_it->second;
+    for (auto& validator : validators) {
+      // A validator can either return false, or throw an EnvoyException.
+      // Both will result in this method throwing an EnvoyException.
+      if (!validator->validate(server_, resources)) {
+        throw EnvoyException("External validator rejected the config.");
```
Comment (htuch) [inline source/common/config/external_config_validators_impl.cc:36]: Test for this specific exception message (same below).

**Review by htuch — COMMENTED** [review]
API and internal interfaces LGTM. Can we get a non-Googler to also sanity check this @envoyproxy/api-shepherds? Thanks.

**Comment by htuch** [comment]
LMK when this is ready for full review @adisuissa.

File: source/common/config/external_config_validators_impl.cc (hunk 6 above)
Comment (adisuissa) [reply source/common/config/external_config_validators_impl.cc:36]: Added some tests.

File: source/common/config/external_config_validators.h (hunk 5 above)
Comment (adisuissa) [reply source/common/config/external_config_validators.h]: I've updated the comment here, let me know if you think it should be further clarified.

File: envoy/upstream/cluster_manager.h (hunk 4 above)
Comment (adisuissa) [reply envoy/upstream/cluster_manager.h:273]: Done.

**Comment by adisuissa** [comment]
@htuch Added some more tests, and it is ready for review.
This PR is becoming pretty large (mainly due to tests), so feel free to tell me to split it.

File: source/common/config/external_config_validators.h (hunk 5 above)
Comment (adisuissa) [reply source/common/config/external_config_validators.h]: Changed the name to Custom Config Validators.

**Comment by adisuissa** [comment]
@htuch should be good to go, PTAL!

**Comment by adisuissa** [comment]
> API and internal interfaces LGTM. Can we get a non-Googler to also sanity check this @envoyproxy/api-shepherds? Thanks.

cc @mattklein123 @lizan for API non-Googler review

File: api/envoy/config/core/v3/config_source.proto [hunk 7]
```diff
@@ -108,6 +110,16 @@ message ApiConfigSource {
...
+
+  // A list of config validators that will be executed when a new update is
+  // received from the ApiConfigSource. Note that each validator handles a
+  // specific xDS service type, and only the validators corresponding to the
+  // type url (in `:ref: DiscoveryResponse` or `:ref: DeltaDiscoveryResponse`)
+  // will be invoked.
+  // If the validator returns false or throws an exception, the config will be rejected by
+  // the client, and a NACK will be sent.
+  // [#extension-category: envoy.config.validators]
+  repeated TypedExtensionConfig config_validators_typed_configs = 9;
```
Comment (lizan) [inline api/envoy/config/core/v3/config_source.proto:122]:
```suggestion
  repeated TypedExtensionConfig config_validators = 9;
```

File: api/envoy/config/core/v3/config_source.proto [hunk 8]
```diff
@@ -240,3 +252,32 @@ message ConfigSource {
   // turn expect to be delivered.
   ApiVersion resource_api_version = 6 [(validate.rules).enum = {defined_only: true}];
 }
+
```
Comment (lizan) [inline api/envoy/config/core/v3/config_source.proto:255]: Why this move?

Comment (adisuissa) [reply api/envoy/config/core/v3/config_source.proto:255]: `ExtensionConfigSource` was defined in *api/envoy/config/core/v3/extension.proto* and uses `ConfigSource` (which is defined in this file).
Therefore if we import *api/envoy/config/core/v3/extension.proto* (so we can use its `TypedExtensionConfig`) in this file we create a cyclic dependency.

Two alternatives:

1. Create a new message, similar to `TypedExtensionConfig` that will be used only in this file.
2. Move `ExtensionConfigSource` from *api/envoy/config/core/v3/extension.proto* to a new file.

I think the current approach is the better than the 2 alternatives, but LMK if you think otherwise.

Comment (lizan) [reply api/envoy/config/core/v3/config_source.proto:255]: SGTM, just wanted to make sure this doesn't break API compatibility.

**Review by lizan — COMMENTED** [review]
/lgtm api

**Comment by adisuissa** [comment]
@htuch anything else that you think we should do here?

**Comment by rojkov** [comment]
@htuch is on vacation until Tue-Wed.

File: envoy/upstream/cluster_manager.h [hunk 9]
```diff
@@ -36,6 +36,12 @@
 #include "absl/container/node_hash_map.h"
 
 namespace Envoy {
+
+// Forward declaration.
+namespace Server {
+class Instance;
```
Comment (htuch) [inline envoy/upstream/cluster_manager.h:42]: Why is this needed?

File: envoy/config/config_validator.h [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+class ConfigValidator {
+public:
+  virtual ~ConfigValidator() = default;
+
+  /**
+   * Validates a given set of resources matching a State-of-the-World update.
+   * @param server A server instance to fetch the state before applying the config.
+   * @param resources List of decoded resources that reflect the new state.
+   * @throw optional EnvoyException if the config should be rejected.
+   * @return whether the config is valid or not.
```
Comment (htuch) [inline envoy/config/config_validator.h:34]: Why do we have both a `false` return option and an `EnvoyException` option to reject?

File: source/extensions/config/validators/minimum_clusters/minimum_clusters_validator.cc [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+      (cur_clusters_num - removed_resources_size >= min_clusters_num_)) {
+    return true;
+  }
+
+  // It could be that the removed clusters gets us below the threshold, simulate what happens if
+  // the current clusters list is updated.
+  uint32_t newly_added_clusters_num = 0;
+  absl::flat_hash_set<std::string> added_cluster_names(added_resources.size());
+  for (const auto& resource : added_resources) {
+    envoy::config::cluster::v3::Cluster cluster =
```
Comment (htuch) [inline source/extensions/config/validators/minimum_clusters/minimum_clusters_validator.cc:51]: Nit: copy

File: source/extensions/config/validators/minimum_clusters/minimum_clusters_validator.cc [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+      ++newly_added_clusters_num;
+    }
+  }
+
+  // Count the clusters that need to be removed.
+  uint32_t removed_clusters_num = 0;
+  for (const auto& removed_cluster : removed_resources) {
+    Upstream::ClusterConstOptRef cluster = cur_clusters.getCluster(removed_cluster);
+    // Only clusters that were added via api can be removed.
+    if ((cluster.has_value()) && (cluster->get().info()->addedViaApi())) {
```
Comment (htuch) [inline source/extensions/config/validators/minimum_clusters/minimum_clusters_validator.cc:69]: Nit: gratuitous parens

File: test/extensions/config/validators/minimum_clusters/config_test.cc [hunk 13]
```diff
@@ -0,0 +1,52 @@
+#include "envoy/extensions/config/validators/minimum_clusters/v3/minimum_clusters.pb.h"
```
Comment (htuch) [inline test/extensions/config/validators/minimum_clusters/config_test.cc:1]: Is there some test to check if we have multiple validators that later ones will be called?

File: source/extensions/config/validators/minimum_clusters/minimum_clusters_validator.cc [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+  // the current clusters list is updated.
+  uint32_t newly_added_clusters_num = 0;
+  absl::flat_hash_set<std::string> added_cluster_names(added_resources.size());
+  for (const auto& resource : added_resources) {
+    envoy::config::cluster::v3::Cluster cluster =
+        dynamic_cast<const envoy::config::cluster::v3::Cluster&>(resource->resource());
+
+    // If the cluster was already added in the current update, skip this cluster.
+    if (!added_cluster_names.insert(cluster.name()).second) {
+      continue;
```
Comment (htuch) [inline source/extensions/config/validators/minimum_clusters/minimum_clusters_validator.cc:56]: Missing coverage as per https://storage.googleapis.com/envoy-pr/2878971/coverage/source/extensions/config/validators/minimum_clusters/minimum_clusters_validator.cc.gcov.html

**Review by htuch — CHANGES_REQUESTED** [review]
LGTM modulo comments.
/wait

File: source/extensions/config/validators/minimum_clusters/minimum_clusters_validator.cc (hunk 14 above)
Comment (adisuissa) [reply source/extensions/config/validators/minimum_clusters/minimum_clusters_validator.cc:55]: Increased coverage.

File: test/extensions/config/validators/minimum_clusters/config_test.cc (hunk 13 above)
Comment (adisuissa) [reply test/extensions/config/validators/minimum_clusters/config_test.cc:1]: I've got the following unit test that tests 2 generic validators: https://github.com/envoyproxy/envoy/blob/f2260c6c4eb688cacf2d6398c71b22698f4534fd/test/common/config/custom_config_validators_impl_test.cc#L141-L160

Let me know if you want me to add an integration test that actually runs 2 Minimum-Clusters with different thresholds.

File: source/extensions/config/validators/minimum_clusters/minimum_clusters_validator.cc (hunk 12 above)
Comment (adisuissa) [reply source/extensions/config/validators/minimum_clusters/minimum_clusters_validator.cc]: Fixed

File: envoy/config/config_validator.h (hunk 10 above)
Comment (adisuissa) [reply envoy/config/config_validator.h]: As we discussed offline, changed to only throw an exception.

File: envoy/upstream/cluster_manager.h (hunk 9 above)
Comment (adisuissa) [reply envoy/upstream/cluster_manager.h]: Good catch, thanks!
It was needed in a previous iteration of this PR, but not anymore, so I removed it.

**Review by htuch — COMMENTED** [review]
LGTM; one last thing, do we have functional test coverage that ensures that with both the old and new unified muxes, for both SotW and delta, that we correctly invoke the validators?

**Review by htuch — APPROVED** [review]
I spoke with @adisuissa about this offline and the unified mux integration testing is an orthogonal concern, so it's fine to merge this PR and deal with that in followup which is planned.

File: source/extensions/config/validators/minimum_clusters/minimum_clusters_validator.cc [hunk 15]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+void MinimumClustersValidator::validate(
+    const Server::Instance& server,
+    const std::vector<Envoy::Config::DecodedResourcePtr>& added_resources,
+    const Protobuf::RepeatedPtrField<std::string>& removed_resources) {
+  const Upstream::ClusterManager& cm = server.clusterManager();
+  // If the number of clusters after removing all of the clusters in the removed_resources list is
+  // above the threshold, then it is surely a valid config.
+  const Upstream::ClusterManager::ClusterInfoMaps cur_clusters = cm.clusters();
```
Comment (pradeepcrao) [inline source/extensions/config/validators/minimum_clusters/minimum_clusters_validator.cc:37]: Did you mean to make this a const ref?

Comment (adisuissa) [reply source/extensions/config/validators/minimum_clusters/minimum_clusters_validator.cc]: done

**Review by htuch — APPROVED** [review]

**Comment by adisuissa** [comment]
/retest

//test/extensions/filters/http/csrf:csrf_filter_integration_test seems unrelated, and seems to be failing other submitted PRs.
